sources/packages/backend/libs/utilities/src/application-data-comparison/application-data-comparison.ts

```diff
@@ -0,0 +1,169 @@
+import {
+  ApplicationDataChange,
+  ChangeTypes,
+} from "./application-data-comparison.models";
+
+/**
+ * Compares two sets of student application data and returns an array of changes.
+ * Identifies differences between the current and previous
+ * data, such as changes in values, additions, or deletions, and represents
+ * them as an array of {@link ApplicationDataChange} objects.
+ * @param currentData current state of the application data.
+ * @param previousData previous state of the application data.
+ * @returns array of {@link ApplicationDataChange} objects representing the detected changes.
+ */
+export function compareApplicationData(
+  currentData: unknown,
+  previousData: unknown,
+): ApplicationDataChange[] {
+  // Initial object to allow the recursive function to add changes.
+  const initialChanges = new ApplicationDataChange();
+  compareApplicationDataRecursive(currentData, previousData, initialChanges);
+  const [root] = initialChanges.changes;
+  return root?.changes ?? [];
+}
+
+/**
+ * Recursively compares two sets of student application data and detect the differences.
+ * @param currentData current state of the student application data.
+ * @param previousData previous state of the student application data.
+ * @param parentChange parent change to add the new changes detected.
+ * @param options options to support the new change creation.
+ * - `propertyKey`: key of the property that changed.
+ * - `index`: index of the array item that changed.
+ */
+function compareApplicationDataRecursive(
+  currentData: unknown,
+  previousData: unknown,
+  parentChange: ApplicationDataChange,
+  options?: { propertyKey?: string; index?: number },
+): void {
+  // Same data, no need to compare.
+  if (isEqual(currentData, previousData)) {
+    return;
+  }
+  // Property has a null or undefined value but had a value before or vice versa.
+  if (
+    (currentData === null ||
+      currentData === undefined ||
+      previousData === null ||
+      previousData === undefined) &&
+    currentData !== previousData
+  ) {
+    const newValueChange = new ApplicationDataChange({
+      key: options?.propertyKey,
+      newValue: currentData,
+      oldValue: previousData,
+      index: options?.index,
+    });
+    parentChange.changes.push(newValueChange);
+    return;
+  }
+  // Check array items.
+  // previousData is checked also to enforce the type as an array. In case
+  // it is null or undefined it will be handled at the previous scenario.
+  if (Array.isArray(currentData) && Array.isArray(previousData)) {
+    checkArrayChanges(currentData, previousData, parentChange, options);
+    return;
+  }
+  // Check object properties.
+  if (typeof currentData === "object") {
+    checkObjectChanges(currentData, previousData, parentChange, options);
```

Comment:
Inside `checkArrayChanges` and `checkObjectChanges`, is it necessary to check `isEqual` again?

Reply:
Yes, you were right. Removed as per our work session. Nice catch 😉